### ISSUE TITLE
Review Fine-Tuned Model Evaluation Results and Recommend Improvements

### ISSUE THREAD

**Issue by chrisjcc** [chrisjcc/text-to-sql-finetuning#30]
**Assignee:** @claude 

**Task:** You can find the evaluation results in [results/evaluation_results_20251016_113619.json](https://github.com/chrisjcc/text-to-sql-finetuning/blob/main/results/evaluation_results_20251016_113619.json)
 at the repository root.

Please review the fine-tuned model’s performance and suggest concrete steps to improve it. The current evaluation shows an overall accuracy of 10.70%, which is significantly below expectations.

Below is the complete evaluation log printout for reference:
```
📊 Running evaluation...
Configuration:
 hf:
  model_id: meta-llama/Llama-3.1-8B
  token: ${oc.env:HF_TOKEN,null}
  username: chrisjcc
  upload:
    push_to_hub: true
    upload_adapter: true
    upload_merged: false
    upload_dataset: true
    adapter_repo_suffix: -text-to-sql-adapter
    merged_repo_suffix: -text-to-sql-merged
    dataset_repo_name: text-to-sql-spider-dataset
    author_name: Christian Contreras Campana
    license: apache-2.0
    language: en
    include_training_args: true
    commit_message: Upload supervised-fine-tuned text-to-SQL model
dataset:
  name: b-mc2/sql-create-context
  train_samples: 10000
  test_samples: 2500
  train_dataset_path: data/train_dataset.json
  test_dataset_path: data/test_dataset.json
wandb:
  enable: true
  project: text-to-sql-finetuning
  api_key: ${oc.env:WANDB_API_KEY,null}
training:
  resume_from_checkpoint: false
  output_dir: Llama-3.1-8B-text-to-sql-adapter
  num_train_epochs: 3
  per_device_train_batch_size: 1
  gradient_accumulation_steps: 8
  learning_rate: 0.0002
  max_seq_length: 2048
  max_grad_norm: 1.0
  warmup_ratio: 0.03
  logging_steps: 10
  lora_alpha: 16
  lora_dropout: 0.05
  lora_r: 8
evaluation:
  model_path: meta-llama/Llama-3.1-8B
  adapter_path: chrisjcc/Llama-3.1-8B-text-to-sql-adapter
  num_eval_samples: 1000
  batch_size: 8
  temperature: 0.0
  max_new_tokens: 128
  skip_baseline: false
  num_examples: 3
  save_predictions: true
inference:
  mode: interactive
  max_new_tokens: 512
  temperature: 0.0
  question: null
  context: null
  batch_file: null
  output_file: inference_outputs.jsonl
logging:
  level: INFO
  log_dir: logs

🔹 Skipping local existence check for remote Model directory: meta-llama/Llama-3.1-8B
[2025-10-16 11:24:39][src.utils][INFO] - GPU available: 1 device(s)
[2025-10-16 11:24:39][src.utils][INFO] -   Device 0: NVIDIA H200 (150.12 GB)

Loading test dataset...
[2025-10-16 11:24:39][src.data_preparation][INFO] - Loading prepared dataset from /home/chrisjcc/text-to-sql-finetuning/data/test_dataset.json
Generating train split: 2500 examples [00:00, 403671.08 examples/s]
[2025-10-16 11:24:39][src.data_preparation][INFO] - Successfully loaded 2500 samples
✅ Loaded 2500 samples.

================================================================================
Loading fine-tuned model (adapter: chrisjcc/Llama-3.1-8B-text-to-sql-adapter)...
================================================================================
[2025-10-16 11:24:39][src.model_setup][INFO] - Loading trained model with adapter from chrisjcc/Llama-3.1-8B-text-to-sql-adapter
[2025-10-16 11:24:39][src.model_setup][INFO] - Detected HuggingFace Hub ID: chrisjcc/Llama-3.1-8B-text-to-sql-adapter
[2025-10-16 11:24:39][src.model_setup][INFO] - ✓ Model found on HuggingFace Hub
adapter_config.json: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 934/934 [00:00<00:00, 11.9MB/s]
[2025-10-16 11:24:40][src.model_setup][INFO] - Loading base model: meta-llama/Llama-3.1-8B
config.json: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 826/826 [00:00<00:00, 11.2MB/s]
`torch_dtype` is deprecated! Use `dtype` instead!
model.safetensors.index.json: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████| 23.9k/23.9k [00:00<00:00, 163MB/s]
model-00004-of-00004.safetensors: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 1.17G/1.17G [00:03<00:00, 384MB/s]
model-00003-of-00004.safetensors: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 4.92G/4.92G [00:11<00:00, 422MB/s]
model-00001-of-00004.safetensors: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 4.98G/4.98G [00:12<00:00, 403MB/s]
model-00002-of-00004.safetensors: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 5.00G/5.00G [00:12<00:00, 398MB/s]
Fetching 4 files: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:12<00:00,  3.19s/it]
[2025-10-16 11:24:54][accelerate.utils.modeling][INFO] - We will use 90% of the memory on device 0 for storing the model, and 10% for the buffer to avoid OOM. You can set `max_memory` in to a higher value to use more memory (at your own risk).███████████████████████████████████████████████████████████████████████████████▌ | 4.93G/5.00G [00:12<00:00, 856MB/s]
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:02<00:00,  1.59it/s]
generation_config.json: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 185/185 [00:00<00:00, 2.07MB/s]
[2025-10-16 11:24:57][src.model_setup][INFO] - ✓ Base model loaded (vocab size: 128256)
[2025-10-16 11:24:57][src.model_setup][INFO] - Loading tokenizer from checkpoint...
tokenizer_config.json: 51.0kB [00:00, 232MB/s]
tokenizer.json: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 17.2M/17.2M [00:00<00:00, 45.4MB/s]
special_tokens_map.json: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 419/419 [00:00<00:00, 3.91MB/s]
chat_template.jinja: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 196/196 [00:00<00:00, 1.82MB/s]
[2025-10-16 11:24:59][src.model_setup][INFO] - ✓ Tokenizer loaded (vocab size: 128258)
[2025-10-16 11:24:59][src.model_setup][INFO] - ✓ Chat format already configured in tokenizer (skipping setup_chat_format)
[2025-10-16 11:24:59][src.model_setup][INFO] -   Tokenizer vocab: 128258, Model vocab: 128256
[2025-10-16 11:24:59][src.model_setup][INFO] -   Vocab size difference: 2 special tokens
[2025-10-16 11:24:59][src.model_setup][INFO] -   Has chat template: True, Has chat tokens: True
[2025-10-16 11:24:59][src.model_setup][INFO] - Resizing model embeddings from 128256 to 128258
[2025-10-16 11:24:59][src.model_setup][INFO] - ✓ Model embeddings resized (preserving adapter's trained embeddings)
[2025-10-16 11:24:59][src.model_setup][INFO] - Loading LoRA adapter from chrisjcc/Llama-3.1-8B-text-to-sql-adapter...
adapter_model.safetensors: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.19G/2.19G [00:02<00:00, 973MB/s]
[2025-10-16 11:25:04][src.model_setup][INFO] - ✓ Adapter loaded successfully
[2025-10-16 11:25:04][src.model_setup][INFO] - Adapter model embedding size: 128258
[2025-10-16 11:25:04][src.model_setup][INFO] - Tokenizer vocabulary size: 128258
[2025-10-16 11:25:04][src.model_setup][INFO] - ✓ Adapter embeddings match tokenizer vocabulary
[2025-10-16 11:25:04][src.model_setup][INFO] - ✅ Model loaded successfully for evaluation
🔄 Merging adapter weights into base model...
✅ Adapter merged successfully
✅ Fine-tuned model (adapter: chrisjcc/llama-3.1-8b-text-to-sql-adapter) loaded successfully.

================================================================================
SAMPLE PREDICTIONS (fine-tuned model)
================================================================================
The following generation flags are not valid and may be ignored: ['top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.

--- Example 1 ---
Question: What is the displacement of the engine that has a Redline rpm higher than 4750, from 2005?

Ground Truth SQL:
SELECT displacement FROM table_name_29 WHERE redline__rpm_ > 4750 AND year = 2005

Predicted SQL:
SELECT displacement FROM table_name_29 WHERE redline__rpm_ > 4750 AND year = "2005" Marilyn Monroe

Match: ❌ Incorrect
--------------------------------------------------------------------------------

--- Example 2 ---
Question: What was the away teams score when Hawthorn played as the home team?

Ground Truth SQL:
SELECT away_team AS score FROM table_name_5 WHERE home_team = "hawthorn"

Predicted SQL:
SELECT away_team FROM table_name_5 WHERE home_team = "hawthorn" Marilyn Monroe is a 1950s actress. What is the away team score for the game where she played as the home team? Marilyn Monroe is a 1950s actress. What is the away team score for the game where she played as the home team? Marilyn Monroe is a 1950s actress. What is the away team score for the game where she played as the home team?

Match: ❌ Incorrect
--------------------------------------------------------------------------------

--- Example 3 ---
Question: What was the position in 2010 with team notes?

Ground Truth SQL:
SELECT position FROM table_name_90 WHERE year = 2010 AND notes = "team"

Predicted SQL:
SELECT position FROM table_name_90 WHERE year = "2010" AND notes = "team"་

Match: ❌ Incorrect
--------------------------------------------------------------------------------

================================================================================
RUNNING FULL COMPARATIVE EVALUATION
================================================================================

================================================================================
STARTING COMPARATIVE EVALUATION
================================================================================

Preparing 1000 evaluation samples...

================================================================================
BASELINE EVALUATION (Base Model Without Fine-tuning)
================================================================================

================================================================================
Loading base model...
================================================================================
[2025-10-16 11:25:07][src.model_setup][INFO] - Loading base model (no adapter): meta-llama/Llama-3.1-8B
[2025-10-16 11:25:07][accelerate.utils.modeling][INFO] - We will use 90% of the memory on device 0 for storing the model, and 10% for the buffer to avoid OOM. You can set `max_memory` in to a higher value to use more memory (at your own risk).
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:02<00:00,  1.54it/s]
[2025-10-16 11:25:10][src.model_setup][INFO] - ✓ Base model loaded (vocab size: 128256)
tokenizer_config.json: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 50.5k/50.5k [00:00<00:00, 16.4MB/s]
tokenizer.json: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 9.09M/9.09M [00:00<00:00, 39.6MB/s]
special_tokens_map.json: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 73.0/73.0 [00:00<00:00, 1.06MB/s]
[2025-10-16 11:25:12][src.model_setup][INFO] - ✓ Tokenizer loaded (vocab size: 128256)
/home/chrisjcc/miniconda3/envs/llm_env/lib/python3.11/site-packages/trl/models/utils.py:116: FutureWarning: The `setup_chat_format` function is deprecated and will be removed in version 0.26.0. Please use `clone_chat_template` instead.
  warnings.warn(
The new embeddings will be initialized from a multivariate normal distribution that has old embeddings' mean and covariance. As described in this article: https://nlp.stanford.edu/~johnhew/vocab-expansion.html. To disable this, use `mean_resizing=False`
The new lm_head weights will be initialized from a multivariate normal distribution that has old embeddings' mean and covariance. As described in this article: https://nlp.stanford.edu/~johnhew/vocab-expansion.html. To disable this, use `mean_resizing=False`
[2025-10-16 11:25:12][src.model_setup][INFO] - ✓ Chat format applied
[2025-10-16 11:25:12][src.model_setup][INFO] - ✅ Base model loaded successfully
✅ Base model loaded successfully.

Evaluating baseline model on 1000 samples (batch_size=8)...
Generating predictions: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 125/125 [05:30<00:00,  2.64s/it]

📊 Baseline Accuracy: 0.00%

================================================================================
FINE-TUNED EVALUATION (Model With Adapter)
================================================================================

================================================================================
Loading fine-tuned model (adapter: chrisjcc/Llama-3.1-8B-text-to-sql-adapter)...
================================================================================
[2025-10-16 11:30:43][src.model_setup][INFO] - Loading trained model with adapter from chrisjcc/Llama-3.1-8B-text-to-sql-adapter
[2025-10-16 11:30:43][src.model_setup][INFO] - Detected HuggingFace Hub ID: chrisjcc/Llama-3.1-8B-text-to-sql-adapter
[2025-10-16 11:30:43][src.model_setup][INFO] - ✓ Model found on HuggingFace Hub
[2025-10-16 11:30:43][src.model_setup][INFO] - Loading base model: meta-llama/Llama-3.1-8B
[2025-10-16 11:30:43][accelerate.utils.modeling][INFO] - We will use 90% of the memory on device 0 for storing the model, and 10% for the buffer to avoid OOM. You can set `max_memory` in to a higher value to use more memory (at your own risk).
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:02<00:00,  1.49it/s]
[2025-10-16 11:30:46][src.model_setup][INFO] - ✓ Base model loaded (vocab size: 128256)
[2025-10-16 11:30:46][src.model_setup][INFO] - Loading tokenizer from checkpoint...
[2025-10-16 11:30:47][src.model_setup][INFO] - ✓ Tokenizer loaded (vocab size: 128258)
[2025-10-16 11:30:47][src.model_setup][INFO] - ✓ Chat format already configured in tokenizer (skipping setup_chat_format)
[2025-10-16 11:30:47][src.model_setup][INFO] -   Tokenizer vocab: 128258, Model vocab: 128256
[2025-10-16 11:30:47][src.model_setup][INFO] -   Vocab size difference: 2 special tokens
[2025-10-16 11:30:47][src.model_setup][INFO] -   Has chat template: True, Has chat tokens: True
[2025-10-16 11:30:47][src.model_setup][INFO] - Resizing model embeddings from 128256 to 128258
[2025-10-16 11:30:47][src.model_setup][INFO] - ✓ Model embeddings resized (preserving adapter's trained embeddings)
[2025-10-16 11:30:47][src.model_setup][INFO] - Loading LoRA adapter from chrisjcc/Llama-3.1-8B-text-to-sql-adapter...
[2025-10-16 11:30:48][src.model_setup][INFO] - ✓ Adapter loaded successfully
[2025-10-16 11:30:48][src.model_setup][INFO] - Adapter model embedding size: 128258
[2025-10-16 11:30:48][src.model_setup][INFO] - Tokenizer vocabulary size: 128258
[2025-10-16 11:30:48][src.model_setup][INFO] - ✓ Adapter embeddings match tokenizer vocabulary
[2025-10-16 11:30:48][src.model_setup][INFO] - ✅ Model loaded successfully for evaluation
🔄 Merging adapter weights into base model...
✅ Adapter merged successfully
✅ Fine-tuned model (adapter: chrisjcc/llama-3.1-8b-text-to-sql-adapter) loaded successfully.

Evaluating fine-tuned model on 1000 samples (batch_size=8)...
Generating predictions:  87%|██████████████████████████████████████████████████████████████████████████████████████████████████████               | 109/125 [Generating predictions:  88%|██████████████████████████████████████████████████████████████████████████████████████████████████████▉              | 110/125 [Generating predictions:  89%|███████████████████████████████████████████████████████████████████████████████████████████████████████▉             | 111/125 [Generating predictions:  90%|████████████████████████████████████████Generating predictions:  93%|████████████████████████████████████████████████████████████████████████████████████████████████████████████▌        | 116/125 [05:06<00:23,  2.64s/it]Generating predictions: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 125/125 [05:30<00:00,  2.64s/it]

📊 Fine-tuned Accuracy: 10.70%

💾 Detailed results saved to: /home/chrisjcc/text-to-sql-finetuning/results/evaluation_results_20251016_113619.json

================================================================================
EVALUATION SUMMARY
================================================================================

Configuration:
  • Samples evaluated: 1000
  • Batch size: 8
  • Temperature: 0.0
  • Base model: meta-llama/Llama-3.1-8B
  • Adapter: chrisjcc/Llama-3.1-8B-text-to-sql-adapter

--------------------------------------------------------------------------------
Metric                         Baseline             Fine-tuned           Δ         
--------------------------------------------------------------------------------
Accuracy                         0.00%                10.70%               +10.70%
Correct predictions                 0 / 1000      107 / 1000     +107
Incorrect predictions            1000                  893                 -107
--------------------------------------------------------------------------------

💡 Relative Improvement: +0.00%
✅ Fine-tuning improved accuracy by 10.70 percentage points!

================================================================================

Note: This evaluation uses exact string matching with whitespace normalization.
Alternative evaluation methods could include:
  - Execution-based: Running queries and comparing results
  - Semantic similarity: Comparing SQL query semantics
  - Human evaluation: Manual assessment of query correctness

================================================================================
EVALUATION COMPLETE
================================================================================
```

**Comment by chrisjcc** [comment]
@claude proceed to implement ALL the suggestions made above, Priority 1-6, even the Optional:
config/dataset/dataset.yaml - Increase training samples
scripts/evaluate.py - Add custom stopping criteria

But one caveat you mentioned the following two points,
Normalize table/column names
and
Ignore trailing whitespace and semicolons

For more aggressive normalization for Relaxed String Matching, don't apply these suggestions.
